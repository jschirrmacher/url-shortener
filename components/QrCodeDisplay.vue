<script setup lang="ts">
interface Props {
  shortCode: string
  size?: number
  showDownload?: boolean
}

const props = withDefaults(defineProps<Props>(), {
  size: 200,
  showDownload: true,
})

const qrCodeUrl = computed(() => `/api/qr/${props.shortCode}?size=${props.size}`)
const downloadUrl = computed(() => `/api/qr/${props.shortCode}?size=${props.size}&download=true`)
const svgDownloadUrl = computed(() => `/api/qr/${props.shortCode}?format=svg&size=${props.size}&download=true`)

const isLoading = ref(true)
const hasError = ref(false)
const errorMessage = ref("")
const isClient = ref(false)

// Check if we're on client side
onMounted(() => {
  isClient.value = true
})

function handleImageLoad() {
  isLoading.value = false
  hasError.value = false
  errorMessage.value = ""
}

function handleImageError() {
  isLoading.value = false
  hasError.value = true
  errorMessage.value = "QR-Code konnte nicht geladen werden"
}

async function downloadQrCode(format: "png" | "svg") {
  try {
    const url = format === "svg" ? svgDownloadUrl.value : downloadUrl.value
    const filename = `qr-${props.shortCode}.${format}`
    const link = document.createElement("a")
    link.href = url
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Download failed:", error)
  }
}

// Reset loading state when shortCode changes
watch(
  () => props.shortCode,
  () => {
    isLoading.value = true
    hasError.value = false
    errorMessage.value = ""
  },
)
</script>

<template>
  <div class="qr-code-container">
    <div class="qr-code-wrapper" :style="{ width: `${size}px`, height: `${size}px` }">
      <!-- Show loading initially -->
      <div v-if="!isClient || isLoading" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
        <div class="text-center">
          <div class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto" />
          <p class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
        </div>
      </div>

      <!-- Show error state -->
      <div
        v-else-if="hasError"
        class="flex items-center justify-center w-full h-full bg-red-50 border-2 border-red-200 rounded-lg"
      >
        <div class="text-center">
          <svg class="mx-auto h-8 w-8 text-red-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.732-.833-2.5 0L4.268 19.5c-.77.833.192 2.5 1.732 2.5z"
            />
          </svg>
          <p class="text-sm text-red-600 mt-1">{{ errorMessage }}</p>
          <p class="text-xs text-gray-500 mt-1">URL: {{ qrCodeUrl }}</p>
        </div>
      </div>

      <ClientOnly>
        <img
          :src="qrCodeUrl"
          :alt="`QR-Code fÃ¼r ${shortCode}`"
          class="w-full h-full object-contain rounded-lg border border-gray-200"
          @load="handleImageLoad"
          @error="handleImageError"
        >
      </ClientOnly>
    </div>

    <!-- Download Buttons - only show when QR code is loaded -->
    <div v-if="isClient && showDownload && !isLoading && !hasError" class="mt-4 flex gap-2 justify-center">
      <button
        class="inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
        @click="downloadQrCode('png')"
      >
        <svg class="h-4 w-4 mr-2" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M12 10v6m0 0l-3-3m3 3l3-3m2 8H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
          />
        </svg>
        PNG herunterladen
      </button>

      <button
        class="inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
        @click="downloadQrCode('svg')"
      >
        <svg class="h-4 w-4 mr-2" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M12 10v6m0 0l-3-3m3 3l3-3m2 8H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
          />
        </svg>
        SVG herunterladen
      </button>
    </div>
  </div>
</template>

<style scoped>
.qr-code-container {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.qr-code-wrapper {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
}
</style>
