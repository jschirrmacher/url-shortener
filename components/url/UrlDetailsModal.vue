<script setup lang="ts">
import { ref, watch, nextTick } from "vue"
import type { UrlRecord, UpdateUrlResponse } from "~/types/index"

interface Props {
  shortCode: string
  shortUrl?: string
  originalUrl?: string
  title?: string
  isOpen: boolean
}

interface Emits {
  close: []
  updated: [url: UrlRecord]
}

const props = defineProps<Props>()
const emit = defineEmits<Emits>()

const dialog = ref<HTMLDialogElement>()
const isDialogMounted = ref(false)

// QR Code state
const qrCodeUrl = ref("")

// Update state
const updateLoading = ref(false)
const updateError = ref("")
const updateSuccess = ref(false)

// Generate QR Code
async function generateQrCode() {
  if (!props.shortCode) return
  
  try {
    // Use the correct API route and get the image directly
    qrCodeUrl.value = `/api/qr/${props.shortCode}?format=png&size=200`
  } catch (err) {
    console.error("Failed to generate QR code:", err)
  }
}

// Download QR Code
async function downloadQrCode(format: 'png' | 'svg') {
  if (!props.shortCode) return
  
  try {
    const response = await fetch(`/api/qr/${props.shortCode}?format=${format}&download=true`)
    const blob = await response.blob()
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `qr-${props.shortCode}.${format}`
    a.click()
    URL.revokeObjectURL(url)
  } catch (err) {
    console.error("Failed to download QR code:", err)
  }
}

// Update URL
async function updateUrl(data: { shortCode: string; originalUrl: string; title: string }) {
  if (!props.shortCode || !data.originalUrl.trim()) return

  try {
    updateLoading.value = true
    updateError.value = ""
    updateSuccess.value = false

    const body: { originalUrl: string; title?: string; newShortCode?: string } = { 
      originalUrl: data.originalUrl.trim() 
    }
    
    if (data.title !== undefined) {
      body.title = data.title.trim()
    }
    
    if (data.shortCode && data.shortCode !== props.shortCode) {
      body.newShortCode = data.shortCode
    }

    const { data: response, error } = await useFetch<UpdateUrlResponse>(`/api/urls/${props.shortCode}`, {
      method: "PUT",
      body,
    })

    if (error.value) {
      throw error.value
    }

    if (response.value?.success && response.value.url) {
      updateSuccess.value = true
      
      const updatedUrl: UrlRecord = {
        shortCode: response.value.url.shortCode,
        originalUrl: response.value.url.originalUrl,
        title: response.value.url.title || "",
        createdAt: new Date().toISOString(),
        createdBy: "current-user",
      }
      
      emit("updated", updatedUrl)
      
      setTimeout(() => {
        closeModal()
      }, 1500)
    }
  } catch (err) {
    const errorObj = err as Error & { data?: { detail?: string } }
    updateError.value = errorObj?.data?.detail ?? errorObj?.message ?? "Fehler beim Aktualisieren der URL"
  } finally {
    updateLoading.value = false
  }
}

function resetUpdate() {
  updateLoading.value = false
  updateError.value = ""
  updateSuccess.value = false
}

function openModal() {
  if (dialog.value && !dialog.value.open) {
    dialog.value.showModal()
    generateQrCode()
    nextTick(() => {
      isDialogMounted.value = true
    })
  }
}

function closeModal() {
  if (dialog.value) {
    dialog.value.close()
    resetUpdate()
    qrCodeUrl.value = ""
    isDialogMounted.value = false
    emit("close")
  }
}

function handleBackdropClick(event: MouseEvent) {
  if (event.target === dialog.value) {
    closeModal()
  }
}

function handleKeydown(event: KeyboardEvent) {
  if (event.key === "Escape") {
    closeModal()
  }
}

// Watch for isOpen prop changes
watch(
  () => props.isOpen,
  (isOpen) => {
    if (isOpen) {
      openModal()
    } else {
      closeModal()
    }
  },
)
</script>

<template>
  <dialog
    ref="dialog"
    class="modal-dialog"
    @click="handleBackdropClick"
    @keydown="handleKeydown"
  >
    <div class="modal-content">
      <ModalHeader 
        :title="`URL Details: ${shortCode}`"
        @close="closeModal"
      />

      <div class="modal-body">
        <QrCodeSection
          :short-code="shortCode"
          :qr-code-url="qrCodeUrl"
          @download="downloadQrCode"
        />

        <AlertMessage
          v-if="updateError"
          type="error"
          title="Fehler beim Aktualisieren"
          :message="updateError"
        />

        <AlertMessage
          v-if="updateSuccess"
          type="success"
          title="Erfolgreich aktualisiert"
          message="Die URL wurde erfolgreich geÃ¤ndert."
        />

        <UrlForm
          :short-url="shortUrl || ''"
          :short-code="shortCode"
          :original-url="originalUrl || ''"
          :title="title"
          @changed="updateUrl"
          @cancel="closeModal"
        />
      </div>
    </div>
  </dialog>
</template>

<style scoped>
.modal-dialog {
  border: none;
  border-radius: 0.5rem;
  box-shadow: 0 20px 25px -5px rgb(0 0 0 / 0.1), 0 8px 10px -6px rgb(0 0 0 / 0.1);
  padding: 0;
  max-width: 32rem;
  width: 100%;
  max-height: 90vh;
}

.modal-dialog::backdrop {
  background-color: rgba(0, 0, 0, 0.5);
}

.modal-content {
  background-color: #ffffff;
  border-radius: 0.5rem;
  max-height: 90vh;
  overflow: hidden;
  display: flex;
  flex-direction: column;
}

.modal-body {
  padding: 1.5rem;
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
  overflow-y: auto;
}
</style>
