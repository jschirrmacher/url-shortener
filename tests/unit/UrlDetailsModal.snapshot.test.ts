/**
 * @vitest-environment happy-dom
 */
import { describe, it, expect, vi, beforeEach } from "vitest"
import { mount } from "@vue/test-utils"
import QrCodeDisplay from "~/components/QrCodeDisplay.vue"
import UrlDetailsModal from "~/components/url/UrlDetailsModal.vue"

describe("URL Components Snapshots", () => {
  describe("QrCodeDisplay", () => {
    it("renders correctly with default props", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "abc123",
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders correctly with custom size", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "abc123",
          size: 300,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders correctly with custom alt text", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "abc123",
          alt: "Custom QR Code Alt Text",
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders correctly with all custom props", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "custom123",
          size: 250,
          alt: "Custom Alt Text",
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders correctly with very long shortCode", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "very-long-short-code-that-might-affect-rendering",
          size: 200,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders correctly with minimal size", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "min123",
          size: 100,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders correctly with large size", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "large123",
          size: 500,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe("UrlDetailsModal", () => {
    beforeEach(() => {
      // Mock global functions
      global.$fetch = vi.fn().mockResolvedValue({
        success: true
      }) as any
    })

    const defaultProps = {
      shortCode: "abc123",
      shortUrl: "http://localhost:3000/abc123",
      originalUrl: "https://example.com",
      isOpen: false
    }

    it("should render when closed", () => {
      const wrapper = mount(UrlDetailsModal, {
        props: defaultProps
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render when open", () => {
      const wrapper = mount(UrlDetailsModal, {
        props: {
          ...defaultProps,
          isOpen: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with long URLs", () => {
      const wrapper = mount(UrlDetailsModal, {
        props: {
          ...defaultProps,
          shortUrl: "http://localhost:3000/long123",
          originalUrl: "https://very-long-example-url-that-should-be-displayed-properly.com/with/many/path/segments",
          isOpen: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with special characters in URLs", () => {
      const wrapper = mount(UrlDetailsModal, {
        props: {
          ...defaultProps,
          shortCode: "special123",
          shortUrl: "http://localhost:3000/special123",
          originalUrl: "https://example.com/path?param=value&other=test#section",
          isOpen: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with different shortCode formats", () => {
      const wrapper = mount(UrlDetailsModal, {
        props: {
          ...defaultProps,
          shortCode: "ABC-123_test",
          shortUrl: "http://localhost:3000/ABC-123_test",
          isOpen: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with minimal props", () => {
      const wrapper = mount(UrlDetailsModal, {
        props: {
          shortCode: "min",
          shortUrl: "http://localhost:3000/min",
          originalUrl: "https://a.co",
          isOpen: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
