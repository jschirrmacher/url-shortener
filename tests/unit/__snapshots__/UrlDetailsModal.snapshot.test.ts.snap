// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`URL Components Snapshots > UrlDetailsModal > should render when closed 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog">
  <div data-v-dd4104c0="" class="modal-content">
    <modalheader data-v-dd4104c0="" title="URL Details: abc123"></modalheader>
    <div data-v-dd4104c0="" class="modal-body">
      <qrcodesection data-v-dd4104c0="" short-code="abc123" qr-code-url="[object Object]"></qrcodesection>
      <alertmessage data-v-dd4104c0="" type="error" title="Fehler beim Aktualisieren" message="[object Object]"></alertmessage>
      <alertmessage data-v-dd4104c0="" type="success" title="Erfolgreich aktualisiert" message="Die URL wurde erfolgreich geändert."></alertmessage>
      <urleditform data-v-dd4104c0="" short-url="http://localhost:3000/abc123" short-code="abc123" original-url="https://example.com" loading="[object Object]"></urleditform>
    </div>
  </div>
</dialog>"
`;

exports[`URL Components Snapshots > UrlDetailsModal > should render when open 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog">
  <div data-v-dd4104c0="" class="modal-content">
    <modalheader data-v-dd4104c0="" title="URL Details: abc123"></modalheader>
    <div data-v-dd4104c0="" class="modal-body">
      <qrcodesection data-v-dd4104c0="" short-code="abc123" qr-code-url="[object Object]"></qrcodesection>
      <alertmessage data-v-dd4104c0="" type="error" title="Fehler beim Aktualisieren" message="[object Object]"></alertmessage>
      <alertmessage data-v-dd4104c0="" type="success" title="Erfolgreich aktualisiert" message="Die URL wurde erfolgreich geändert."></alertmessage>
      <urleditform data-v-dd4104c0="" short-url="http://localhost:3000/abc123" short-code="abc123" original-url="https://example.com" loading="[object Object]"></urleditform>
    </div>
  </div>
</dialog>"
`;

exports[`URL Components Snapshots > UrlDetailsModal > should render with different shortCode formats 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog">
  <div data-v-dd4104c0="" class="modal-content">
    <modalheader data-v-dd4104c0="" title="URL Details: ABC-123_test"></modalheader>
    <div data-v-dd4104c0="" class="modal-body">
      <qrcodesection data-v-dd4104c0="" short-code="ABC-123_test" qr-code-url="[object Object]"></qrcodesection>
      <alertmessage data-v-dd4104c0="" type="error" title="Fehler beim Aktualisieren" message="[object Object]"></alertmessage>
      <alertmessage data-v-dd4104c0="" type="success" title="Erfolgreich aktualisiert" message="Die URL wurde erfolgreich geändert."></alertmessage>
      <urleditform data-v-dd4104c0="" short-url="http://localhost:3000/ABC-123_test" short-code="ABC-123_test" original-url="https://example.com" loading="[object Object]"></urleditform>
    </div>
  </div>
</dialog>"
`;

exports[`URL Components Snapshots > UrlDetailsModal > should render with long URLs 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog">
  <div data-v-dd4104c0="" class="modal-content">
    <modalheader data-v-dd4104c0="" title="URL Details: abc123"></modalheader>
    <div data-v-dd4104c0="" class="modal-body">
      <qrcodesection data-v-dd4104c0="" short-code="abc123" qr-code-url="[object Object]"></qrcodesection>
      <alertmessage data-v-dd4104c0="" type="error" title="Fehler beim Aktualisieren" message="[object Object]"></alertmessage>
      <alertmessage data-v-dd4104c0="" type="success" title="Erfolgreich aktualisiert" message="Die URL wurde erfolgreich geändert."></alertmessage>
      <urleditform data-v-dd4104c0="" short-url="http://localhost:3000/long123" short-code="abc123" original-url="https://very-long-example-url-that-should-be-displayed-properly.com/with/many/path/segments" loading="[object Object]"></urleditform>
    </div>
  </div>
</dialog>"
`;

exports[`URL Components Snapshots > UrlDetailsModal > should render with minimal props 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog">
  <div data-v-dd4104c0="" class="modal-content">
    <modalheader data-v-dd4104c0="" title="URL Details: min"></modalheader>
    <div data-v-dd4104c0="" class="modal-body">
      <qrcodesection data-v-dd4104c0="" short-code="min" qr-code-url="[object Object]"></qrcodesection>
      <alertmessage data-v-dd4104c0="" type="error" title="Fehler beim Aktualisieren" message="[object Object]"></alertmessage>
      <alertmessage data-v-dd4104c0="" type="success" title="Erfolgreich aktualisiert" message="Die URL wurde erfolgreich geändert."></alertmessage>
      <urleditform data-v-dd4104c0="" short-url="http://localhost:3000/min" short-code="min" original-url="https://a.co" loading="[object Object]"></urleditform>
    </div>
  </div>
</dialog>"
`;

exports[`URL Components Snapshots > UrlDetailsModal > should render with special characters in URLs 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog">
  <div data-v-dd4104c0="" class="modal-content">
    <modalheader data-v-dd4104c0="" title="URL Details: special123"></modalheader>
    <div data-v-dd4104c0="" class="modal-body">
      <qrcodesection data-v-dd4104c0="" short-code="special123" qr-code-url="[object Object]"></qrcodesection>
      <alertmessage data-v-dd4104c0="" type="error" title="Fehler beim Aktualisieren" message="[object Object]"></alertmessage>
      <alertmessage data-v-dd4104c0="" type="success" title="Erfolgreich aktualisiert" message="Die URL wurde erfolgreich geändert."></alertmessage>
      <urleditform data-v-dd4104c0="" short-url="http://localhost:3000/special123" short-code="special123" original-url="https://example.com/path?param=value&amp;other=test#section" loading="[object Object]"></urleditform>
    </div>
  </div>
</dialog>"
`;
